src/devices/sound/setapcm.cpp

```diff
@@ -0,0 +1,433 @@
+{
+	switch (reg & 0xf)
+	{
+	case 0x02: // Start position LSB, also affected current position?
+		m_start = m_pos = (m_start & ~mem_mask) | (data & mem_mask);
+		break;
+	case 0x03: // Start position MSB
+		m_start = m_pos = (m_start & ~(u32(mem_mask) << 16)) | (u32(data & mem_mask) << 16);
+		break;
+	case 0x05: // Flags
+		COMBINE_DATA(&m_flags);
+		break;
+	case 0x06: // Frequency
+		COMBINE_DATA(&m_freq);
+		break;
+	case 0x07: // Loop start position LSB
+		m_lpstart = (m_lpstart & ~mem_mask) | (data & mem_mask);
+		break;
+	case 0x09: // Loop start position MSB
+		m_lpstart = (m_lpstart & ~(u32(mem_mask) << 16)) | (u32(data & mem_mask) << 16);
+		break;
+	case 0x0a: // Loop end position LSB
+		m_lpend = (m_lpend & ~mem_mask) | (data & mem_mask);
+		break;
+	case 0x0b: // Loop end position MSB
+		m_lpend = (m_lpend & ~(u32(mem_mask) << 16)) | (u32(data & mem_mask) << 16);
+		break;
+	case 0x0c: // End position LSB
+		m_end = (m_end & ~mem_mask) | (data & mem_mask);
+		break;
+	case 0x0d: // End position MSB
+		m_end = (m_end & ~(u32(mem_mask) << 16)) | (u32(data & mem_mask) << 16);
+		break;
+	case 0x0e: // Right volume
+		COMBINE_DATA(&m_vol_r);
+		break;
+	case 0x0f: // Left volume
+		COMBINE_DATA(&m_vol_l);
+		break;
+	}
+}
+
+//-------------------------------------------------
+//  snd_w - write each voice registers
+//-------------------------------------------------
+
+template<unsigned MaxVoices, unsigned Divider>
+void setapcm_device<MaxVoices, Divider>::snd_w(offs_t offset, u16 data, u16 mem_mask)
+{
+	m_stream->update();
+
+	const int v = (offset >> 4);
+
+	if (v >= MAX_VOICES)
+		return;
+
+	m_voice[v].reg_w(offset & 0xf, data, mem_mask);
+	//logerror("v%02d: %04x & %04x to reg %02d (PC=%x)\n", v, data, mem_mask, r, machine().describe_context());
+}
+
+
+//-------------------------------------------------
+//  snd_r - read each voice registers
+//-------------------------------------------------
+
+template<unsigned MaxVoices, unsigned Divider>
+u16 setapcm_device<MaxVoices, Divider>::snd_r(offs_t offset)
+{
+	const int v = (offset >> 4);
+
+	if (v >= MAX_VOICES)
+		return 0;
+
+	m_stream->update();
+
+	return m_voice[v].reg_r(offset & 0xf);
+}
+
+
+//-------------------------------------------------
+//  key_w - set keyon/off flags for each voices
+//-------------------------------------------------
+
+template<unsigned MaxVoices, unsigned Divider>
+void setapcm_device<MaxVoices, Divider>::key_w(offs_t offset, u16 data, u16 mem_mask)
+{
+	const u16 prev = m_keyctrl;
+
+	m_stream->update();
+
+	COMBINE_DATA(&m_keyctrl);
+
+	//logerror("KEYCTRL: %04x -> %04x %s\n", prev, m_keyctrl, machine().describe_context());
+
+	u16 delta = prev ^ m_keyctrl;
+	for (int v = 0; v < MAX_VOICES; v++)
+	{
+		if (BIT(delta, v))
+		{
+			if (BIT(m_keyctrl, v) && (!(BIT(prev, v)))) // keyon
+			{
+				m_voice[v].keyon();
+			}
+			else if ((!(BIT(m_keyctrl, v))) && BIT(prev, v)) // keyoff
+			{
+				m_voice[v].keyoff();
+			}
+		}
```

Comment:
You’re doing redundant tests here now.  Since you’re already testing that the bit changed (`if (BIT(delta, v))` on line 409), you don’t need to check `prev` at all in the inner `if` statements.  You already know it changed so you can just do `if (BIT(m_keyctrl, v)) { ... } else { ... }`